### PR TITLE
Components callbacks

### PR DESCRIPTION
**Pull request by artem30801**
## About this pull request

Added support of component callbacks!
Rewiews on naming of functions, classes and arguments will be very appreciated, as some of the current method names feel too long or ambigous.

## Changes
- Extracted everything not related to function of slash commands from `CommandObject` to new `CallbackObject` base class
- Added `ComponentCallbackObject` and `CogComponentCallbackObject` to hold component callback data (inherits from `CallbackObject`)
   - Thus, component callbacks support max_concurency, cooldowns, checks and individual error handlers (same as slash commands do now)
- Added methods to `SlashCommand` to add, remove and manage component callbacks
   - `add_component_callback` to registeer function as component callback
      -  Support of lists (including empty lists) of message and custom ids
      - `custom_id` defaults to function name in not specified
   - `extend_component_callback` to dynamically add combination of message_id and custom_id to existing callback object
   - `get_component_callback` to get component callback by message_id, custom_id, component_type
   - `remove_component_callback` to dynamically remove combination of message_id and custom_id from listening for callbacks
   - `remove_component_callback_obj`  to remove all listeners (combinations of message_id and custom_id) assosiated with specified callback object
- Added `component_callback` decorator to `SlashCommand` to register function as component callback
- Added `cog_component` decorator to `cog_ext`  to register cog method as component callback
- Added `on_component_callback` event triggered along with actual component callback
- Added component callback error handling event `on_component_callback_error` to work in same manner as slash commands do now

## Checklist

- [X] I've checked this pull request runs on `Python 3.6.X`.
- [ ] This fixes something in [Issues](https://github.com/eunwoo1104/discord-py-slash-command/issues).
    - Issue:
- [X] This adds something new.
- [X] There is/are breaking change(s).
- [X] (If required) Relevant documentation has been updated/added.
- [ ] This is not a code change. (README, docs, etc.)

### CODE REVIEW

File: discord_slash/client.py [hunk 1]
```diff
@@ -847,6 +879,230 @@ def wrapper(cmd):
...
+        :type use_callback_name: bool
+        :param component_type: The type of the component to avoid collisions with other component types. See :class:`.model.ComponentType`.
+        :type component_type: Optional[int]
+        :raises: .error.DuplicateCustomID, .error.IncorrectFormat
+        """
+        if message_id and message_ids is not None:
+            raise error.IncorrectFormat("You cannot use both `message_id` and `message_ids`!")
+
+        if custom_id and custom_ids is not None:
+            raise error.IncorrectFormat("You cannot use both `custom_id` and `custom_ids`!")
```
Comment (LordOfPolls) [inline discord_slash/client.py]: Would it not be more user friendly to have just `custom_ids` and have it handle `str` and `list[str]`

Comment (artem30801) [reply discord_slash/client.py]: Maybe, I just was trying to be consistent with stuff like `embed` and `embeds`

Comment (artem30801) [reply discord_slash/client.py]: Changed all signatures of callback adding functions to include suggested changes. They also now accept whole message objects and component dicts (or lists of those too)

**Review by LordOfPolls — APPROVED** [review]

File: discord_slash/client.py [hunk 2]
```diff
@@ -847,6 +880,203 @@ def wrapper(cmd):
...
+        use_callback_name=True,
+        component_type: int = None,
+    ):
+        """
+        Adds a coroutine callback to a component.
+        Callback can be made to only accept component interactions from a specific messages
+        and/or custom_ids of components.
+
+        :param callback: The coroutine to be called when the component is interacted with. Must accept a single argument with the type :class:`.context.ComponentContext`.
+        :type callback: Coroutine
```
Comment (LilSpazJoekp) [inline discord_slash/client.py]:
```suggestion
        :param Coroutine callback: The coroutine to be called when the component is interacted with. Must accept a single argument with the type :class:`.context.ComponentContext`.
```
These can be simplified to be on one line.

File: discord_slash/client.py [hunk 3]
```diff
@@ -847,6 +880,203 @@ def wrapper(cmd):
...
+    def component_callback(
+        self,
+        *,
+        messages: typing.Union[int, discord.Message, list] = None,
+        components: typing.Union[str, dict, list] = None,
+        use_callback_name=True,
+        component_type: int = None,
+    ):
+        """
+        Decorator that registers a coroutine as a component callback.\n
```
Comment (LilSpazJoekp) [inline discord_slash/client.py]:
```suggestion
        Decorator that registers a coroutine as a component callback.
```

File: discord_slash/error.py [hunk 4]
```diff
@@ -40,6 +40,20 @@ def __init__(self, name: str):
         super().__init__(f"Duplicate command name detected: {name}")
 
 
+class DuplicateCallback(SlashCommandError):
+    """
+    There is a duplicate component custom ID.
+    """
```
Comment (LilSpazJoekp) [inline discord_slash/error.py]:
```suggestion
    """Indicate exceptions where there is a duplicate component custom ID."""
```

**Review by LilSpazJoekp — COMMENTED** [review]
Just a few suggestions.

File: discord_slash/error.py (hunk 4 above)
Comment (artem30801) [reply discord_slash/error.py]: I don't whink it's consistent with other exceptions docstings, like:
-     Some formats are incorrect. See Discord API DOCS for proper format.
-     There is a duplicate command name.
-     Request to Discord API has failed.
-     There are duplicate :class:`.SlashCommand` instances.

Comment (LilSpazJoekp) [reply discord_slash/error.py]: What about 
```suggestion
    """Raised when there is a duplicate component custom ID."""
```

Comment (i0bs) [reply discord_slash/error.py]: I'm not sure I follow, they're only docstrings in this case. Are we trying to strictly follow PEP-8 here?

**Review by i0bs — APPROVED** [review]
LGTM

File: discord_slash/error.py (hunk 4 above)
Comment (LordOfPolls) [reply discord_slash/error.py]: Yeah im gunna say this isnt worth delaying merge, feel free to discuss in staff chats. and possibly commit later

**Pull request merged**